Firestore/Swift/Tests/Codable/CodableDocumentTests.swift

```diff
@@ -0,0 +1,550 @@
+        try container.encode(id, forKey: .id)
+        var nestedContainer = container.nestedContainer(keyedBy: NestedCodingKeys.self, forKey: .nested)
+        try nestedContainer.encode(name, forKey: .name)
+      }
+    }
+
+    let model = Model(id: 12345, name: "ModelName")
+    let dict = ["id": 12345,
+                "nested": ["name": "ModelName"]] as [String: Any]
+
+    assertRoundTrip(model: model, encoded: dict)
+  }
+
+  class SuperModel: Codable, Equatable {
+    var superPower: Double? = 100.0
+    var superName: String? = "superName"
+
+    init(power: Double, name: String) {
+      superPower = power
+      superName = name
+    }
+
+    static func == (lhs: SuperModel, rhs: SuperModel) -> Bool {
+      return (lhs.superName == rhs.superName) && (lhs.superPower == rhs.superPower)
+    }
+
+    private enum CodingKeys: String, CodingKey {
+      case superPower
+      case superName
+    }
+
+    required init(from decoder: Decoder) throws {
+      let container = try decoder.container(keyedBy: CodingKeys.self)
+      superPower = try container.decode(Double.self, forKey: .superPower)
+      superName = try container.decode(String.self, forKey: .superName)
+    }
+
+    func encode(to encoder: Encoder) throws {
+      var container = encoder.container(keyedBy: CodingKeys.self)
+      try container.encode(superPower, forKey: .superPower)
+      try container.encode(superName, forKey: .superName)
+    }
+  }
+
+  class SubModel: SuperModel {
+    var timestamp: Timestamp? = Timestamp(seconds: 848_483_737, nanoseconds: 23423)
+
+    init(power: Double, name: String, seconds: Int64, nano: Int32) {
+      super.init(power: power, name: name)
+      timestamp = Timestamp(seconds: seconds, nanoseconds: nano)
+    }
+
+    static func == (lhs: SubModel, rhs: SubModel) -> Bool {
+      return ((lhs as SuperModel) == (rhs as SuperModel)) && (lhs.timestamp == rhs.timestamp)
+    }
+
+    private enum CodingKeys: String, CodingKey {
+      case timestamp
+    }
+
+    required init(from decoder: Decoder) throws {
+      let container = try decoder.container(keyedBy: CodingKeys.self)
+      timestamp = try container.decode(Timestamp.self, forKey: .timestamp)
+      try super.init(from: container.superDecoder())
+    }
+
+    override func encode(to encoder: Encoder) throws {
+      var container = encoder.container(keyedBy: CodingKeys.self)
+      try container.encode(timestamp, forKey: .timestamp)
+      try super.encode(to: container.superEncoder())
+    }
+  }
+
+  func testClassHierarchy() {
+    let model = SubModel(power: 100, name: "name", seconds: 123_456_789, nano: 654_321)
+    let dict = ["super": ["superPower": 100, "superName": "name"],
+                "timestamp": Timestamp(seconds: 123_456_789, nanoseconds: 654_321)] as [String: Any]
+
+    assertRoundTrip(model: model, encoded: dict)
+  }
+
+  func testEcodingEncodableArrayNotSupported() {
+    struct Model: Codable, Equatable {
```

Comment:
Does this actually need to be enforced by the Encoder? I can see someone using the Encoder directly to incrementally build things up into a larger value. Enforcing this here prevents doing that kind of thing with arrays.

This is already enforced at the point where we parse the value in `FSTUserDataConverter`.

Reply:
This is because the return type is [String:Any], there is no meaningful way to encode an array into that.

We could add an overload to return [[String:Any]] instead. Is that what you have in mind?